austation/code/modules/mining/minebot_upgrades.dm

```diff
@@ -0,0 +1,269 @@
+#define MBU_TRACKS_B /obj/item/minebot_upgrade/tracks_b
+#define MBU_GUN_A /obj/item/minebot_upgrade/gun_a
+#define MBU_GUN_B /obj/item/minebot_upgrade/gun_b
+#define MBU_GUN_C /obj/item/minebot_upgrade/gun_c
+#define MBU_GUN_D /obj/item/minebot_upgrade/gun_d
+#define MBU_MED /obj/item/minebot_upgrade/
+#define MBU_SIGHT /obj/item/minebot_upgrade/
+#define MBU_LAMP /obj/item/minebot_upgrade/
+#define MBU_NAME /obj/item/minebot_upgrade/
+#define MBU_APC /obj/item/minebot_upgrade/
+#define GUN_UPGRADES list(MBU_GUN_A, MBU_GUN_B)
+#define GUNTYPES list(MBU_GUN_C, MBU_GUN_D)
+
+
+/////////////////////////  PROCS  /////////////////////////
+
+
+/obj/item/minebot_upgrade  //  can't use mine_bot_upgrade because existing definitions conflict with what we want to do
+	name = "minebot upgrade"
+	icon_state = "door_electronics"
+	icon = 'icons/obj/module.dmi'
+	var/weight = 0  //  how many mod slots the upgrade takes up.  each minebot can only install mods until the total weight is 3
+	var/upgrade_id = null  //  populate with the DEFINES above, they point to the class of the upgrade
+	var/list/incompatible = list()  //  this list blacklists new mods from being added (such as gun upgrades for weapons you do not have equipped)
+
+/obj/item/minebot_upgrade/proc/is_minebotgun()
+	if(upgrade_id in GUNTYPES)
+		return TRUE
+
+/obj/item/minebot_upgrade/proc/is_minebotgun_upg()
+	if(upgrade_id in GUN_UPGRADES)
+		return TRUE
+
+/obj/item/minebot_upgrade/afterattack(mob/living/simple_animal/hostile/mining_drone/M, mob/user, proximity)
+	. = ..()
+	if(!istype(M) || !proximity || !upgrade_id)  //  something has gone very wrong.  the user probably wouldn't understand a warning for these
+		return
+	if(upgrade_id in M.upgrades_list)  //  can't install the same mod twice
+		to_chat(user, "<span class='warning'[M] already has a [src] installed!</span>")
+		return
+	if(M.mod_total >= 3  && !(upgrade_id in GUNTYPES))  //  we allow GUNTYPES because they replace an existing mod, instead of adding another.
+		to_chat(user, "<span class='warning'[M] has no more room to install a [src]!</span>")
+		return
+	for(var/obj/item/minebot_upgrade/U in M.upgrades_list)
+		if(upgrade_id in U.incompatible)
+			to_chat(user, "<span class='warning'there are conflicting mods already installed to \the [M]!  You may remove them with a crowbar.</span>")
+			return
+	src.upgrade_bot(M, user)
+
+/obj/item/minebot_upgrade/proc/upgrade_bot(mob/living/simple_animal/hostile/mining_drone/M, mob/user)
+	to_chat(user, "<span class='notice'you install a [src] upgrade into \the [M]</span>")
+	M.upgrades_list += upgrade_id
+	M.mod_total += src.weight
+	qdel(src)
+
+/obj/item/minebot_upgrade/proc/uninstall(mob/living/simple_animal/hostile/mining_drone/M, mob/user)
+	if(!get_turf(M))
+		if(is_minebotgun())
+			to_chat(user, "<span class='warning'there is no room to uninstall the [M]'s previous gun.</span>")
+		else
+			to_chat(user, "<span class='warning'there is no room to uninstall the [src].</span>")
```

Comment:
```suggestion
		if(is_minebotgun())
			to_chat(user, "<span class='warning'>there is no room to uninstall [is_minebotgun() ? \the [M]'s previous gun." : \the [src]]</span>")
```

Reply:
I didn't know you could do that

Reply:
I really liked this idea, but it broke the build and I couldn't work out the correct way to implement it.
I may try to do it this way later, once the content is finished.

Reply:
@Ohagi-Chan My bad, fixed it. I accidently put a `"` in the middle 😅

Reply:
> I didn't know you could do that

http://www.byond.com/docs/ref/info.html#/operator/%3f
😎